### PR TITLE
[incubator/jaeger] Improve the Jaeger chart

### PR DESCRIPTION
**Pull request by pavelnikolov**
This PR adds the following improvements to the *Jaeger* chart:
* Use ConfigMap for configurable parameters
* Add the Hotrod example app
* Allow only some of the components to be installed
* Add support for the spark dependencies job (as a [k8s cronjob](https://kubernetes.io/docs/concepts/workloads/controllers/cron-jobs/)) 
* Use `provisionDataStore` key in the `values.yaml` file instead of `tags` to configure data store provisioning
* Refactor chart to remove unnecessary quotes
* Remove the command overrides of the docker images and use [environment variables configuration](http://jaeger.readthedocs.io/en/latest/deployment/#configuration) instead
* Fix hard-coded replica count
* Collector service works both as `NodePort` and `ClusterIP` service types

### CODE REVIEW

**Comment by pavelnikolov** [comment]
Thanks to @mikelorant for helping me with this PR.

**Comment by pavelnikolov** [comment]
FYI @dvonthenen @jpkrohling @yurishkuro @pavolloffay

**Comment by foxish** [comment]
ok to test

**Comment by foxish** [comment]
Unassigning myself - don't know enough about Jaeger to approve/review.

File: incubator/jaeger/values.yaml
Comment (pavolloffay) [inline incubator/jaeger/values.yaml]: It should run just before the end of the day. Let say 5-10 min for startup

**Comment by pavolloffay** [comment]
does it expose `zipkin` service by default (9411 port to collect spans)?

**Comment by pavelnikolov** [comment]
@pavolloffay 
**TL;DR** yes, it does.

 Take a look at the `values.yaml` file line [#L128](https://github.com/pavelnikolov/charts/blob/dac7a59d6685a47bdc7551e811f4670e0d41d387/incubator/jaeger/values.yaml#L128) - this is where the default port is set. Then, if you take a look at the `collector-deploy.yaml` file [#L83](https://github.com/pavelnikolov/charts/blob/dac7a59d6685a47bdc7551e811f4670e0d41d387/incubator/jaeger/templates/collector-deploy.yaml#L83), you will notice that the same container port is used. Also, on [#L73](https://github.com/pavelnikolov/charts/blob/dac7a59d6685a47bdc7551e811f4670e0d41d387/incubator/jaeger/templates/collector-deploy.yaml#L73), the `COLLECTOR_ZIPKIN_HTTP_PORT` environment variable is configured, which in turn gets its value from the `collector.zipkin.http-port
 ` config map key. Finally, on [#20](https://github.com/pavelnikolov/charts/blob/dac7a59d6685a47bdc7551e811f4670e0d41d387/incubator/jaeger/templates/common-cm.yaml#L20) in the `common-cm.yaml` file you can see that the value comes from the default value (i.e. `9411` from the `values.yaml`).

**Comment by davidvonthenen** [comment]
Will take a look this afternoon as well. Been out of the office.

**Comment by davidvonthenen** [comment]
Giving this a test drive today...

Test Cases:
 - [x] New Cassandra instance
 - [x] New ES instance
 - [x] Existing Cassandra instance
 - [x] Existing ES instance

Will be adding a few more...

**Comment by davidvonthenen** [comment]
@pavolloffay is there any reason why we need both sets of parameters?

```
provisionDataStore:
  cassandra: true
  elasticsearch: false

storage:
  # allowed values (cassandra, elasticsearch)
  type: cassandra
```

It is possible to replace `.Values.storage.type` using `.Values.provisionDataStore.cassandra` for the if statements instead? The reason why I ask it because I ran into a split brain (two different parameters are controlling the storage deploy and configuration) where I didnt know storage.type existed/needed-to-be-changed but I had enabled provisionDataStore.elasticsearch=true. By having only a single parameter, it's a little more straight forward. The default is currently Cassandra, but that can be changed to ES. My understanding is (I could be wrong) that ES is prefered storage type by the Jaeger docs/maintainers.

**Comment by pavelnikolov** [comment]
@dvonthenen The only reason there are two sets of `elasticsearch` and `cassandra` params is for clarity in the requirements conditions. Probably `provisionDataStore` can be turned into a boolean then we can get rid of the duplication.

**Comment by davidvonthenen** [comment]
@pavelnikolov I think it would go a long wait in terms of usability to consolidate. Minimally, there needs to be an update to the README to update the example command line in the `Installing the Chart using a New ElasticSearch Cluster` section to add the `--set storage.type=elasticsearch` command line parameter. That is what led to this discussion. Will finish off the verification today.

File: incubator/jaeger/templates/common-cm.yaml
Comment (davidvonthenen) [inline incubator/jaeger/templates/common-cm.yaml]: This forces the external or previously deployed cassandra host to be <relname>-cassandra. This should be set to  `.Values.storage.cassandra.host`.

File: incubator/jaeger/templates/common-cm.yaml
Comment (davidvonthenen) [inline incubator/jaeger/templates/common-cm.yaml]: This forces the external or previously deployed elasticsearch host to be <relname>-elasticsearch. This should be set to  `.Values.storage.elasticsearch.host`.

**Review by davidvonthenen — COMMENTED** [review]
Found a couple of issues when trying to deploy the chart using previously deployed cassandra and elasticsearch stores.

File: incubator/jaeger/templates/common-cm.yaml
Comment (pavelnikolov) [reply incubator/jaeger/templates/common-cm.yaml]: What values are you using it with? I have tested this with both external and provisioned ES and it is working fine for me. Take a closer look at this helper:
```
{{- define "elasticsearch.client.url" -}}
{{- $port := .Values.storage.elasticsearch.port | toString -}}
{{- if .Values.provisionDataStore.elasticsearch -}}
{{- $host := printf "%s-%s-%s" .Release.Name "elasticsearch" "client" | trunc 63 | trimSuffix "-" -}}
{{- printf "%s://%s:%s" .Values.storage.elasticsearch.scheme $host $port }}
{{- else }}
{{- printf "%s://%s:%s" .Values.storage.elasticsearch.scheme .Values.storage.elasticsearch.host $port }}
{{- end -}}
{{- end -}}
```
It doesn't force anything, right?

Comment (davidvonthenen) [reply incubator/jaeger/templates/common-cm.yaml]: Sorry.... small goof. I got hit by the `storage.type` thing again by not setting it. I corrected that now but it looks like the collector and query components are having problems connecting to elasticsearch.

These are the command line parameters I am running with:
```
--name myrel --set provisionDataStore.cassandra=false --set provisionDataStore.elasticsearch=false --set storage.type=elasticsearch --set storage.elasticsearch.host=elasticsearch --set storage.elasticsearch.port=9200 --set storage.elasticsearch.user=elastic --set storage.elasticsearch.password=changeme
```

Get pods output:
```
[dev@k8controller ~]$ kubectl get pods
NAME                                     READY     STATUS             RESTARTS   AGE
elasticsearch-0                          1/1       Running            0          24m
elasticsearch-1                          1/1       Running            0          24m
elasticsearch-2                          1/1       Running            0          24m
myrel-jaeger-agent-31jxv                 1/1       Running            0          8m
myrel-jaeger-agent-89kxf                 1/1       Running            0          8m
myrel-jaeger-agent-gldbw                 1/1       Running            0          8m
myrel-jaeger-agent-lh8tx                 1/1       Running            0          8m
myrel-jaeger-agent-t7c0h                 1/1       Running            0          8m
myrel-jaeger-agent-wgvrm                 1/1       Running            0          8m
myrel-jaeger-collector-623630575-b32fr   0/1       CrashLoopBackOff   6          8m
myrel-jaeger-query-3159756439-nk7c6      0/1       CrashLoopBackOff   6          8m
```

Error:
```
{"level":"fatal","ts":1515108228.394109,"caller":"collector/main.go:99","msg":"Unable to set up builder","error":"health check timeout: no Elasticsearch node available"
```

Verified the configmap has the correct value:
```
Name:		myrel-jaeger
Namespace:	default
Labels:		app=jaeger
		chart=jaeger-0.3.0
		heritage=Tiller
		jaeger-infra=common-configmap
		release=myrel
Annotations:	<none>

Data
====
...
es.server-urls:
----
http://elasticsearch:9200
...
```

File: incubator/jaeger/templates/common-cm.yaml
Comment (davidvonthenen) [reply incubator/jaeger/templates/common-cm.yaml]: Seems like we might need another define in helpers like you did for elasticsearch to toggle between using the `template "cassandra.fullname"` and `.Values.storage.cassandra.host`.

File: incubator/jaeger/templates/common-cm.yaml
Comment (pavelnikolov) [reply incubator/jaeger/templates/common-cm.yaml]: Oh, good catch. I've tested with ES without username and password. I'll add those two if they are set. As you can see in this file on lines 60 and 62 they haven't been implemented.

File: incubator/jaeger/templates/common-cm.yaml
Comment (pavelnikolov) [reply incubator/jaeger/templates/common-cm.yaml]: Good catch! I've tested provisioned cassandra, external ES and provisioned ES. This was the only scenario that I haven't tested. I'll fix it right away. Thank you very much for catching this!

File: incubator/jaeger/templates/common-cm.yaml
Comment (davidvonthenen) [reply incubator/jaeger/templates/common-cm.yaml]: @pavelnikolov I would also recommend setting the default username and password to elastic/changeme since those are the defaults.

File: incubator/jaeger/templates/_helpers.tpl
Comment (pavelnikolov) [inline incubator/jaeger/templates/_helpers.tpl:34]: This opening comment was missing in the previous PR that got merged. I noticed it while I was resolving conflicts.

**Comment by davidvonthenen** [comment]
Will give the remaining test cases a go today.

**Comment by davidvonthenen** [comment]
Finished the testing. Everything checks out now! Awesome work @pavelnikolov!

LGTM!
 
cc: @unguiculus

**Comment by pavelnikolov** [comment]
I'm planning to move credentials to k8s Secret, but I'll do it in another PR, since this is already too big.

File: incubator/jaeger/values.yaml
Comment (pavelnikolov) [reply incubator/jaeger/values.yaml]: I configured it to run every day @ 23:49

**Comment by ledor473** [comment]
I'd like to also this in our environment and report back!

File: incubator/jaeger/templates/collector-svc.yaml
Comment (ledor473) [inline incubator/jaeger/templates/collector-svc.yaml:23]: Is there any reason why we are overriding the NodePort's port?   
By default it will be assigned to something between 30000-32767 (https://kubernetes.io/docs/concepts/services-networking/service/#type-nodeport)

This would mean that only a single Collector can run on the default port per Kubernetes cluster.

File: incubator/jaeger/values.yaml
Comment (ledor473) [inline incubator/jaeger/values.yaml:7]: Seems like it was mentioned but never changed. I feel that we should have a single `storage.provisionDataStore` instead of 2 booleans

File: incubator/jaeger/README.md
Comment (ledor473) [inline incubator/jaeger/README.md]: The new `storage.**` parameters are not listed here

File: incubator/jaeger/values.yaml
Comment (ledor473) [inline incubator/jaeger/values.yaml]: Seems like `useNodePort` is not used anywhere else

**Review by ledor473 — COMMENTED** [review]
Great update to the chart!
Right now, I've just looked at the changes but I'll test it shortly and report back any issue

File: incubator/jaeger/values.yaml
Comment (davidvonthenen) [reply incubator/jaeger/values.yaml:7]: The reason why you need both booleans is because you need to be able to represent 4 states.
Provision a Cassandra instance
Provision a ElasticSearch instance
Use an existing Cassandra instance
Use an existing ElasticSearch instance

Comment (ledor473) [reply incubator/jaeger/values.yaml:7]: Totally! But that could be done using the following:

a) Provision a Cassandra instance
```
storage: 
  type: cassandra
  provisionDataStore: true
```

b) Provision a ElasticSearch instance
```
storage: 
  type: elasticsearch
  provisionDataStore: true
```

c) Use an existing Cassandra instance
```
storage: 
  type: cassandra
  provisionDataStore: false
  cassandra:
    host: cassandra
    ....
```

d) Use an existing ElasticSearch instance
```
storage: 
  type: elasticsearch
  provisionDataStore: false
  elasticsearch: 
    host: elastic
    ...
```

Comment (davidvonthenen) [reply incubator/jaeger/values.yaml:7]: Agreed. It can be done that way... it just complicates the requirements.yaml to match the condition for the subchart logic. No strong opinion either way.

File: incubator/jaeger/templates/collector-deploy.yaml
Comment (ledor473) [inline incubator/jaeger/templates/collector-deploy.yaml:39]: I think this should be `SPAN_STORAGE` but I could not confirm that during my testing..
https://github.com/jaegertracing/jaeger/blob/master/plugin/storage/factory_config.go#L24-L30

File: incubator/jaeger/values.yaml
Comment (pavelnikolov) [reply incubator/jaeger/values.yaml:7]: As I said earlier I'd be happy to do it this way, if someone shows me what the expression would be in the `requirements.yaml`.

File: incubator/jaeger/README.md
Comment (pavelnikolov) [reply incubator/jaeger/README.md]: I'll fix it shortly

File: incubator/jaeger/templates/collector-svc.yaml
Comment (pavelnikolov) [reply incubator/jaeger/templates/collector-svc.yaml:23]: The reason we need this is for cases when an agent is on one cluster and the collector is on another and we need a predefined port. We have multiple K8s clusters and each node on each cluster has the agent, but only one of our clusters has a collector which ingests all the tracing data.

**Review by ledor473 — COMMENTED** [review]
I've tested it today and it seems to work correctly 👍 

It also fix the same problem that my PR is fixing (different way but they both work) https://github.com/kubernetes/charts/pull/3100

Important to note that the default `type` of the Query Service is now `ClusterIP` so I had to change that as well

File: incubator/jaeger/templates/collector-svc.yaml
Comment (ledor473) [reply incubator/jaeger/templates/collector-svc.yaml:23]: Gotcha! On our side, we are using `type: LoadBalancer` for that. 👍

File: incubator/jaeger/templates/collector-deploy.yaml
Comment (pavelnikolov) [reply incubator/jaeger/templates/collector-deploy.yaml:39]: Run the `hotrod` app to check that it works fine. They are both accepted. If you run `docker run -it --rm jaegertracing/jaeger-collector /go/bin/collector-linux -h`, you'll see that the collector accepts the following flag: `--span-storage.type`. This also automatically means that it would get the value from the uppercased variable with the same name where `-` and `.` are replaced with `_` (i.e. `SPAN_STORAGE_TYPE`). This is mentioned in [the documentation](http://jaeger.readthedocs.io/en/latest/deployment/#configuration).

**Review by egeland — APPROVED** [review]

File: incubator/jaeger/templates/collector-svc.yaml
Comment (mikelorant) [reply incubator/jaeger/templates/collector-svc.yaml:23]: This means that if you are sending tracing in cluster, you are going to create a load balancer. This is totally unnecessary for an in cluster solution and ends up costing money.

File: incubator/jaeger/templates/collector-deploy.yaml
Comment (yurishkuro) [reply incubator/jaeger/templates/collector-deploy.yaml:39]: NB: https://github.com/jaegertracing/jaeger/issues/642

**Comment by pavelnikolov** [comment]
Any more comments?

File: incubator/jaeger/templates/collector-svc.yaml
Comment (ledor473) [reply incubator/jaeger/templates/collector-svc.yaml:23]: No, a Service with type LoadBalancer also has a ClusterIP that can be reached from within the cluster directly (no outbound traffic at all) 👍

File: incubator/jaeger/templates/collector-deploy.yaml
Comment (yurishkuro) [reply incubator/jaeger/templates/collector-deploy.yaml:39]: I merged https://github.com/jaegertracing/jaeger/issues/642, the env var is back to `SPAN_STORAGE_TYPE` (will be released in 1.2)

**Comment by ledor473** [comment]
@pavelnikolov For me, it's good ✅  

My PR https://github.com/kubernetes/charts/pull/3100 was merged so you will probably have to rebase/update your branch

**Comment by pavelnikolov** [comment]
Just rebased.

File: incubator/jaeger/templates/collector-deploy.yaml
Comment (pavelnikolov) [inline incubator/jaeger/templates/collector-deploy.yaml:100]: @ledor473 I used the named ports from your PR, but the values are no longer hard-coded. Now they come from the `values.yaml` and are mapped in the config map.

File: incubator/jaeger/README.md
Comment (pavelnikolov) [reply incubator/jaeger/README.md]: done

File: incubator/jaeger/templates/collector-deploy.yaml
Comment (ledor473) [reply incubator/jaeger/templates/collector-deploy.yaml:100]: @pavelnikolov yes that is even better considering the Collector Pod will listen on that port instead of doing port-forwarding at the Service level 👍

**Comment by pavelnikolov** [comment]
Any ideas why the `tide` check is still pending?

**Comment by pavelnikolov** [comment]
@unguiculus any changes required? Did you want me to assign this PR to someone else? Or break it into smaller pieces?

File: incubator/jaeger/templates/_helpers.tpl
Comment (unguiculus) [inline incubator/jaeger/templates/_helpers.tpl]: This won't work once 1.10 is out. Also, this expects alpha features to be enabled for < 1.8. Better use `Capabilities.APIVersions.Has`.

**Review by unguiculus — CHANGES_REQUESTED** [review]

File: incubator/jaeger/templates/_helpers.tpl
Comment (pavelnikolov) [reply incubator/jaeger/templates/_helpers.tpl]: Done. I didn't know this was possible! I just learned something new! 🎉

Comment (pavelnikolov) [reply incubator/jaeger/templates/_helpers.tpl]: As for:
> Also, this expects alpha features to be enabled for...

This has been documented in the **Prerequisites** section of the `README.md` file and the spark cronjob is disabled by default.

**Comment by mattfarina** [comment]
@dvonthenen there have been some small changes to it since you said LGTM. Still LGTM?

**Comment by davidvonthenen** [comment]
Will check it out

**Comment by davidvonthenen** [comment]
LGTM

File: incubator/jaeger/templates/collector-deploy.yaml
Comment (mattfarina) [inline incubator/jaeger/templates/collector-deploy.yaml:2]: A quick note, Deployments as extensions/v1beta1 will be removed from Kubernetes soon. Since this there has been `apps/v1beta1`, `apps/v1beta2`, and `apps/v1` API versions. For broad compatibility we'd like to see charts use `apps/v1beta2` for the moment. In k8s 1.10 the API versions prior to that can be removed.

This is just a note. Not going to have it hold up this pull request.

Comment (pavelnikolov) [reply incubator/jaeger/templates/collector-deploy.yaml:2]: Thank you! I'll make sure I update it when 1.10 is out.

Comment (pavelnikolov) [reply incubator/jaeger/templates/collector-deploy.yaml:2]: @mattfarina should I remove `extensions/v1beta1` now that k8s 1.10 is out?

**Pull request merged**
